ACES2 Output Transforms

remia opened the pull request:
Opening this PR early even though it's very much a work in progress / prototype so far, the code is more or less imported directly from the CTL.

Next steps includes:
* Adding all the official output transforms from the ampas repository
* Adding support for creative white point and surround gamma factor
* Review lookup table constexpr initialization (assuming OCIO 2.4 moves to C++14, C++14 has limitations in constexpr support making it less clean, increase library size (probably), custom compiler flags needed to lift constexpr instruction count limit), implement caching for custom output transform's tables
* Review current CPU code duplication (reimplement NPM derivation, matrix inversion, etc) that are available in other place in OCIO, though typically using the Ops abstraction, not native code / types
* Optimize CPU path (currently 8x slower than ACES 1)
* Implement GPU path
* Cleanup the FixedFunctions used for testing
* Add tonescale (Daniele's curve) only FixedFunction

doug-walker commented:
Thank you so much for doing this Remi!  Everything aside from ACES2CPUHelpers.h looks perfect. And please assume that OCIO will move to C++14 minimum if that allows a cleaner implementation.

I understand, as you wrote above, that the ACES2CPUHelpers.h module is a direct port of the CTL. But wow, this still really feels like research code. I'm surprised it's only 8x slower than ACES 1, there are so many unnecessary conversions being done. I know you wrote above that optimization is still a TODO, but it really seems like that is going to be a huge task in order to get something reasonably performant. I don't think it's simply a matter of pre-multiplying a few matrices and swapping in the fast power function. For example, I see things like:
```
    const float linear = Hellwig_J_to_Y(inputJMh[0], inputJMhParams) / reference_luminance;
    const float luminanceTS = tonescale_fwd(linear, params);
    const float tonemappedJ = Y_to_Hellwig_J(luminanceTS, inputJMhParams);
```
And that seems like it should be replaced by a spline or something. But building these approximations would take time and may not be exact. I'm very worried about the amount of work here, given that OCIO 2.4 is supposed to release in only a few months.

remia commented:
Pushed an update with my current work in progress, this mainly split the new FixedFunctionOps into 3 for JMh conversion, Tonescale / ChromaCompress and GamutCompress. The code still follows the CTL from aces-dev and has not been optimized further yet (as discussed).

One point that will probably be interesting to agree on is the FixedFunctionOps split strategy, as this will be hard to change later and might impact the optimization. There is still some duplication of work here and there (like Matrix that could be merged) and the GPU path also has texture duplication where ChromaCompression and GamutCompression uses the same tables.

KelSolaar commented:
Awesome @remia! I built some test wheels on test.pypi.org: https://test.pypi.org/project/opencolorio/

The Windows tests were breaking so I bypassed them: https://github.com/AcademySoftwareFoundation/OpenColorIO/compare/main...colour-science:OpenColorIO:aces2-output-transforms

KelSolaar commented:
Looking at the builtins, we have those currently:

```
ACES-2-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-VIDEO-REC709lim
ACES-2-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-VIDEO-REC709-D60lim
ACES-2-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-VIDEO-1000nit-P3D65lim
```

It would be useful for me to have all of them, even if they are null ops so that we can move on the config side. I'm becoming mildly concerned about the 15 September deadline considering the quantity of work left.

remia commented:
@KelSolaar I added more (if not all) transforms in my latest commit and started a [spreadsheet](https://docs.google.com/spreadsheets/d/1BxfwNGeL90tWj-ZI-lytHmw3b_cwYcCnHRJc3kkzQTw/edit?gid=0#gid=0) to list everything I have so far. I'm not opposed to changing the list or names if any feedback comes. Note that I dropped the distinction between VIDEO and CINEMA as there is not really any difference I think so far (eg. no surround compensation in the CTL).

doug-walker commented:
Thanks for the updates Remi! This looks really good. I have a few comments:
1. It might be useful to reorder the D60 tab in the spreadsheet to match the D65 one so that it's easier to compare one to the other, but they do seem to match up (each D65 transform has a matching D60 version).
2. The Built-in transform names are all fine with me except for the use of "XYZ-E" in some of the D60 ones. I understand that on the aces-dev side this means that XYZ is normalized to illuminant E. However, we've been using a different convention in OCIO. In OCIO, XYZ is ALWAYS normalized to illuminant E, and so the dash is used to indicate the adaptive white point. So I would propose changing those to "XYZ-D60", or if you don't like that, then just "XYZ".
3. You wrote that you added "more (if not all)" the transforms. Comparing to Scott's list, it seems like you have all of them (25 each for D65 and D60). Are there others you are thinking of adding? Thomas seems to have 50 output transforms in his spreadsheet too.
4. @KelSolaar do you have what you need to proceed with the config generator?

num3ric commented:
Once fixed, I'm also getting these additional warnings + error in HLSL:
```
Shader@0x000001AD29D520E0(266,23-61): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
Shader@0x000001AD29D520E0(281,47-90): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
Shader@0x000001AD29D520E0(282,28-75): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
Shader@0x000001AD29D520E0(302,19-38): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
Shader@0x000001AD29D520E0(303,22-47): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
Shader@0x000001AD29D520E0(65,18-109): warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop
Shader@0x000001AD29D520E0(63,3-25): error X3511: unable to unroll loop, loop does not appear to terminate in a timely manner (1024 iterations)
```

KelSolaar commented:
Just built new wheels, will check the builtins soon: https://test.pypi.org/project/opencolorio/2.4.0.dev2/

KelSolaar commented:
Can see a bunch of *stuff* in the builtins :)

```
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-REC709_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-108nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-300nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-500nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-1000nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-2000nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-4000nit-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-500nit-REC2020_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-1000nit-REC2020_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-2000nit-REC2020_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-4000nit-REC2020_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-REC709-D60-in-REC709-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-REC709-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-REC709-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-P3-D60-in-XYZ-E_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-108nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-300nit-P3-D60-in-XYZ-E_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-500nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-1000nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-2000nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-4000nit-P3-D60-in-P3-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-500nit-P3-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-1000nit-P3-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-2000nit-P3-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-4000nit-P3-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-500nit-REC2020-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-1000nit-REC2020-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-2000nit-REC2020-D60-in-REC2020-D65_2.0',
'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - HDR-4000nit-REC2020-D60-in-REC2020-D65_2.0',
```

remia commented:
Thanks for the feedback!

> It might be useful to reorder the D60 tab in the spreadsheet to match the D65 one so that it's easier to compare one to the other, but they do seem to match up (each D65 transform has a matching D60 version).

I just did a re-order pass, now both tabs should be aligned and sorted by CTL names.

> The Built-in transform names are all fine with me except for the use of "XYZ-E" in some of the D60 ones. I understand that on the aces-dev side this means that XYZ is normalized to illuminant E. However, we've been using a different convention in OCIO. In OCIO, XYZ is ALWAYS normalized to illuminant E, and so the dash is used to indicate the adaptive white point. So I would propose changing those to "XYZ-D60", or if you don't like that, then just "XYZ".

Based on the other names, especially the D60 ones, the part after the "in" is meant to be the encoding primaries (of the container space), which are needed for the white scaling step. For example `'ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-P3-D60-in-P3-D65_2.0'`, so I think it best we use just XYZ rather than XYZ-D60 in this case. What do you think?

> You wrote that you added "more (if not all)" the transforms. Comparing to Scott's list, it seems like you have all of them (25 each for D65 and D60). Are there others you are thinking of adding? Thomas seems to have 50 output transforms in his spreadsheet too.

I think everything is here, sorry for the misleading wording!

> The error stems from while (i_lo + 1 < i_hi) in ocio_gamut_cusp_table1_sample. This could potentially be fixed by inserting a [loop] instruction, limiting support to SM4 and above I believe. Related note being that OCIO's HLSL enum shouldn't use _DX11, but rather a shader model version: separate CL though.
> 
> EDIT: Actually, [loop] would not work here since the loop contains texture fetches, see: warning X3570: gradient instruction used in a loop with varying iteration, attempting to unroll the loop. It seems like the function may need to be rewritten in a different way here?

Thanks for sharing those warnings / errors, that's really helpful as I don't have an easy way to check HLSL.

I agree the texture sampling in the while loop is currently far from ideal (probably even from a performance only standpoint). In theory I think there would not be a need to compute derivative here as we are really doing NEAREST sampling on the texture, not needing to pick a minification or magnification filter (but there might very well be other reasons I'm unaware of), in my GLSL tests I used texelFetch() for example, but to maintain compatibility with older implementations we can't rely on that, I assume HLSL may have similar features.

I'm still working on more comprehensive comparison against CTLs on actual images, seeing some texture sampling issues as well on some transforms that might be related.

nick-shaw commented:
> I think it best we use just XYZ rather than XYZ-D60 in this case.

I also think it would be preferable not to include "D60" in the name as that could be confusing. In the ACES repo, the white point used in the name is the parameter value for encoding white point, so is the value that needs to be passed to the [`display_encoding()`](https://github.com/ampas/aces-output/blob/4b693455727501f5b020de1189658a6ed6f5fa9a/d60/dcdm/Output.Academy.P3-D60_48nit_in_XYZ-E_Gamma2pt6.ctl#L103) function. For DCDM, [Illuminant E](https://github.com/ampas/aces-output/blob/4b693455727501f5b020de1189658a6ed6f5fa9a/d60/dcdm/Output.Academy.P3-D60_48nit_in_XYZ-E_Gamma2pt6.ctl#L32) needs to be passed. If you use "D60" in the name, somebody could assume that is the value which should be passed to match your result.

KelSolaar commented:
I started to fill-up the spreadsheet: https://docs.google.com/spreadsheets/d/1z3xsy3sF0I-8AN_tkMOEjHlAs13ba7VAVhrE8v4WIyo/edit?gid=1761235937#gid=1761235937

This will require cross-checking but I think that we are missing quite a few builtins and I could not assign some of them:

- `ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-REC709-D60-in-P3-D65_2.0`
- `ACES-OUTPUT - ACES2065-1_to_CIE-XYZ-D65 - SDR-100nit-P3-D60-in-XYZ-E_2.0`

I haven't had a chance to check the code yet to find out what they are. That being said, the assignment/filling is not exactly trivial as the name are not always close.

remia commented:
Hi @KelSolaar,

The assignment of ViewTransform to CTLs should be available in the [spreadsheet](https://docs.google.com/spreadsheets/d/1BxfwNGeL90tWj-ZI-lytHmw3b_cwYcCnHRJc3kkzQTw/edit?gid=0#gid=0) shared earlier. I agree the names are not directly matching in some cases, for example all the 48 nits use the same parameters as the 100 nits, an alternative would be to duplicate the ViewTransform in OCIO to have them under different names but not sure I'd like that.

KelSolaar commented:
> Hi @KelSolaar,
> 
> The assignment of ViewTransform to CTLs should be available in the [spreadsheet](https://docs.google.com/spreadsheets/d/1BxfwNGeL90tWj-ZI-lytHmw3b_cwYcCnHRJc3kkzQTw/edit?gid=0#gid=0) shared earlier. I agree the names are not directly matching in some cases, for example all the 48 nits use the same parameters as the 100 nits, an alternative would be to duplicate the ViewTransform in OCIO to have them under different names but not sure I'd like that.

Ah thanks! I actually thought about the spreadsheet after posting, it was late and I was intending to check today.

Thanks for the 48 nits explanation, that makes sense to me, I will adjust our other spreadsheets accordingly and see how far I was :D

remia commented:
I pushed an update that should fix the HLSL shader compilation issue by extracting the hue component from the texture into a constant array, as suggested by @KevinJW. Also included some minor updates to keep up with the CTL code (though this will have to be more thoroughly reviewed).

Also sharing a [script](https://gist.github.com/remia/2466fe140d6a4a73a10d1dd461044f14) and OCIO config I have been using to compare CTL and OCIO results on images. This is a very simple script so far and is mostly valuable for the CTL to OCIO display / view mapping with the accompanying OCIO test config. I updated the reference CTL to use the parametric chroma norm.

From what I can tell, here are a list of remaining TODOs before merging in the hope to share the workload:
* Agree on the public APIs changes (set of new exposed Builtins and FixedFunction). As discussed yesterday, some choices made for the FixedFunction splitting might impact later optimisation work so it could be useful to have a way to hide the new FixedFunction for the time being
* Agree on the use of ongoing simplification work not yet present in the CTL reference (chromaCompress parametric norm is currently in OCIO but not in the official CTL, Kevin's implementation potential improvements)
* Review of the current CPU and GPU code and potential cleanup of remaining redundant computation
* Further validation against the CTL reference implementation on patterns and test images with an appropriate error metric, I got reasonable match using a simple image like Kodak LAD but expect edge cases to show up in an image like ACES syntheticChart
* Validation on HLSL

num3ric commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -347,6 +349,1070 @@ void Add_GamutComp_13_Inv_Shader(GpuShaderText & ss,
+    else
+    {
+        ss.newLine() << ss.floatDecl("lo") << " = " << ss.sampleTex2D(name, "vec2((i_lo + 0.5f) / 360.f, 0.5f)") << ".r;";
+        ss.newLine() << ss.floatDecl("hi") << " = " << ss.sampleTex2D(name, "vec2((i_hi + 0.5f) / 360.f, 0.5f)") << ".r;";
```
Sorry I left a previously unpublished review comment here: `float2Const` should be used here instead of `vec2` (same elsewhere).

remia replied:
Thanks @num3ric good catch, I fixed this in the latest commit.

KelSolaar commented:
@remia: As of 6585f18c889d68f236548ae536811a0b27571458, I can only see the 31 builtins [from the comment above](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1983#issuecomment-2295180317). Is this expected or should I have the 54 from the spreadsheet? I haven't had time to check the code to see what is going on.

remia commented:
> @remia: As of https://github.com/AcademySoftwareFoundation/OpenColorIO/commit/6585f18c889d68f236548ae536811a0b27571458, I can only see the 31 builtins https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1983#issuecomment-2295180317. Is this expected or should I have the 54 from the spreadsheet? I haven't had time to check the code to see what is going on.

Yes this is correct, multiple CTLs can share the same Builtin transform if only the display encoding changes. I added a count of unique values below the tables in the spreadsheet, it indeed sum to 31.

KelSolaar commented:
Thanks, I just checked too, I have an issue in the generator, I need to look at what is going on there instead.

remia commented:
Good catch, I fixed that now, hopefully should work!

KevinJW commented:
Just pulled the branch and I get a new faiilure that I didn't previously ...

[ctest] ERROR: test_style (FixedFunctionTransformTest.FixedFunctionTransformTest.test_style)
[ctest] Test the setStyle() and getStyle() methods.
[ctest] ----------------------------------------------------------------------
[ctest] Traceback (most recent call last):
[ctest]   File "C:\Users\kwheatle\source\OpenColorIO\tests\python\FixedFunctionTransformTest.py", line 64, in test_style
[ctest]     self.tr.setStyle(style)
[ctest] PyOpenColorIO.PyOpenColorIO.Exception: 16known FixedFunction transform style:

remia commented on tests/cpu/Config_tests.cpp:
```diff
@@ -4922,6 +4944,189 @@ OCIO_ADD_TEST(Config, fixed_function_serialization)
+        const std::string strEnd =
+            "    from_scene_reference: !<GroupTransform>\n"
+            "      children:\n"
+            "        - !<FixedFunctionTransform> {style: ACES2_OutputTransform, params: []}\n";
```
This triggered a crash on the Windows CI, which I couldn't replicate anywhere including my own Windows workstation. But clang sanitizer caught the issue:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/vector:1538:12: runtime error: reference binding to null pointer of type 'double'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/vector:1538:12 in 
```

Added a check in [OCIOYaml.cpp](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1983/files#diff-efc8a26ff824080bff3b85e2605aa47b120d2fa14b0e8a586b447c6405b3a62dR1354).

zachlewis reviewed (approved):
Remi, this is truly amazing, incredibly thorough work. 

I think you made excellent design decisions, especially considering the complexity of the work. In the end, I think opting for fewer FixedFunctions is the correct decision; and I really love the parameterization -- peak luminance + output chromaticities + nothing else, which makes building higher level functions around 'em much more intuitive. In a config, it's nice to see an array of numbers that actually feels super human-friendly, which is not something I thought I'd find myself saying about the interfaces to ACES2 stuff.

I've been messing with this PR in python for the past few hours -- everything behaves exactly as expected. I can't really speak to the efficiency of the implementations, other than to say stuff feels fast, and the generated shader code looks smart. 

Really nice work, man. No notes, LGTM!

remia commented on src/OpenColorIO/transforms/builtins/ACES.cpp:
```diff
@@ -1047,6 +1114,326 @@ void RegisterAll(BuiltinTransformRegistryImpl & registry) noexcept
+            P3_D60::primaries,
+            CIE_XYZ_ILLUM_E::primaries,
+            1.f,
+            false
```
Good points @zachlewis, however I believe it is as intended and following the [aces-output](https://github.com/ampas/aces-output/tree/main/d60/dcdm) reference. My understanding is that the DCDM SDR standard account for a range of whitepoint at peak luminance through the normalization constant so the white scale is not needed (eg. for D60). DCDM HDR however use a straight PQ encoding so would benefit from the white scaling to preserve the whitepoint at peak luminance.

@nick-shaw is that correct?

nick-shaw replied:
That is indeed correct. That is why the DCDM OTs have a `linear_scale_factor` of 48/52.37 to produce 48 nits on a DCI projector with 52.37 nit peak white for headroom.

remia replied:
Thanks Nick. Noting that in the OCIO implementation, the `linear_scale_factor` is applied as part of the Builtins transforms only when the intent was to have peak luminance scaling (eg. for the 108 and 300 nits HDR transforms). When `linear_scale_factor` is used for the DCDM scaling constant, I chose to ignore that and let the OCIO display handle it.

nick-shaw replied:
Yes. I didn't follow the code path through enough (on my phone) to see how you handled that. But I assumed you did. The DCDM output effectively produces a result normalised to 48 nits, which may have values above 1.0. Then that gets put in a 52.37 nit 'container'.

zachlewis replied:
Excellent, that makes perfect sense.

doug-walker replied:
I agree that it should be false. And I agree that the 48/52.37 scaling should be left to the OCIO display color space.

However, I'm wondering if the D60 PQ versions should be false as well? They are true in the CTL, so Remi has correctly implemented that, but the reasoning is less clear.

The purpose of the Scale White option seems to be to darken values slightly to avoid clipping when the white moves to be non-equal RGB. This is necessary if the encoding does not go above 1., but PQ has plenty of headroom (much more than DCDM), so it would seem unnecessary.  @nick-shaw , I guess this is done out of a worry about QC rejecting masters that exceed a certain nit level in any channel?

nick-shaw replied:
I raised that exact question during a meeting regarding the headroom inherent in PQ. The decision was made to scale anyway to keep all channels within range, as you say.

doug-walker replied:
By "headroom" I was just referring to the fact that PQ encodes at least up to 10,000 nits, so there is no need to darken the 1000, 2000, or 4000 nit versions to fit. In other words, I'm assuming for the D60 1000 nit version without scaling, that the G channel would be roughly at 1000 and the red channel would be somewhat above that, but would not clip. But I think what Nick is confirming is that QC might fail a master where any channel goes above 1000, even if nothing is clipped and even if the luminance is technically not above 1000, and so that was the argument for darkening it.

remia replied:
Thanks for the explanation Doug, agree that it will not clip in the encoding, though I'm not sure about downstream whether there is a [constant] luminance check or if some devices or further conversion steps could simply naïvely clip per channel. Seems good to be on the safe side I guess!

nick-shaw replied:
> I'm assuming for the D60 1000 nit version without scaling, that the G channel would be roughly at 1000 and the red channel would be somewhat above that, but would not clip.

Correct. If you want the maths for a 1000 nit example:
```
>>> D60_peak_xyY = [0.32168, 0.33767, 1000.0]
>>> D60_peak_XYZ = xyY_to_XYZ(D60_peak_xyY)
>>> vecmul(XYZ_to_P3D65_matrix, D60_peak_XYZ)
array([ 1037.77164583,   996.28782762,   923.30520654])
```
So we end up dividing the values by 1.03777 to keep all channels <= 1000. This does mean the actual luminance of 1000 nit D60 in D65 is 1000 / 1.03777 = ~963.6 nits.

doug-walker commented on src/OpenColorIO/transforms/builtins/ACES.cpp:
```diff
@@ -497,6 +497,73 @@ void Generate_roll_white_d65_ops(OpRcPtrVec & ops)
+
+        MatrixOpData::MatrixArrayPtr matrixToXYZ
+            = build_conversion_matrix_to_XYZ_D65(limiting_pri, ADAPTATION_NONE);
+        CreateMatrixOp(ops, matrixToXYZ, TRANSFORM_DIR_FORWARD);
```
Although the limiting_pri is sometimes D60, it does seem that the ADAPTATION_NONE is the correct option to use to match what the CTL is doing for the current set of output transforms. But I am puzzling over it.

All of the current output transforms fall in two sets: either D65, or D60 in a D65 container. Unlike ACES 1, aside from the DCDM, there are no displays that are not calibrated to D65. However, if there was a desire to add a D60 calibrated display (or some other white point), I was trying to think how that would work. The way we have been doing the built-in ACES configs, the non-D65 displays have a Bradford adaptation from the D65 of the reference space to their calibration white. 

I think we could use the D65 view transforms for a rendition that will be at the display's calibration white point but if we want to provide a true D65 or D60 version, we would need new view transforms. Those view transforms may need to undo the Bradford adaptation in the display.

remia replied:
It seems to me the concept of "white point simulation" does not map easily to OCIO v2 viewing transforms workflows. I'm not sure if there is an ideal solution with the current system. As a further option, maybe one could also create a dedicated display for the D60 Sim in D65 case where the display side processing would chromatically adapt from D65 to D60, then apply a white scaling factor like is done in the builtin currently. This is assuming the builtin always "correctly" apply a chromatic adaptation to D65 as intended by cie_xyz_d65. This should allow to avoid duplicated view transforms?

doug-walker reviewed (approved):
I've completed a more detailed pass through this awesome work. Thank you so much Remi for helping us find a path up the very steep mountain that is ACES 2!

remia commented on src/OpenColorIO/ops/fixedfunction/ACES2/Transform.cpp:
```diff
@@ -0,0 +1,928 @@
+float Y_to_J(float Y, const JMhParams &params)
+{
+    float F_L_Y = pow(params.F_L * std::abs(Y) / reference_luminance, 0.42f);
+    return std::copysign(1.f, Y) * reference_luminance * pow(((400.f * F_L_Y) / (27.13f + F_L_Y)) / params.A_w_J, surround[1] * params.z);
```
Will fix the pow to powf, the [reference](https://en.cppreference.com/w/cpp/numeric/math/pow) is a bit confusing as to wether or not pow always mean double argument or not, I assume there is a templated version that accept float but will update for consistency. I will try to get some final CPU profiling but yes pow is definitively the bottleneck at the moment from what I saw before.

remia commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -347,6 +349,1305 @@ void Add_GamutComp_13_Inv_Shader(GpuShaderText & ss,
+    ss.newLine() << "lms = " << "lms * " << ss.float3Const(p.D_RGB[0], p.D_RGB[1], p.D_RGB[2]) << ";";
+
+    ss.newLine() << ss.float3Decl("F_L_v") << " = pow(" << p.F_L << " * abs(lms) / 100.f, " << ss.float3Const(0.42f) << ");";
+    ss.newLine() << ss.float3Decl("rgb_a") << " = (400.f * sign(lms) * F_L_v) / (27.13f + F_L_v);";
```
No, I'll remove them, just checked again, float 32 bits is the default type for literal in GLSL / HLSL.

remia reviewed (commented):
Thanks for the review @doug-walker! I'll have a look at all your points in detail, but answering some of them already to keep discussion going.

doug-walker commented on tests/cpu/ops/fixedfunction/FixedFunctionOpCPU_tests.cpp:
```diff
@@ -405,7 +406,518 @@ OCIO_ADD_TEST(FixedFunctionOpCPU, aces_gamut_map_13)
+    std::vector<float> input_32f(num_samples * num_channels, 0.f);
+    std::vector<float> output_32f(num_samples * num_channels, 0.f);
+
+    GenerateIdentityLut3D(input_32f.data(), lut_size, num_channels, OCIO::LUT3DORDER_FAST_RED);
```
I actually missed the fact that the test does inv, then fwd, so please ignore my comment about trying to improve the spacing with a gamma. But you are right that perhaps the input should be scaled down so that you don't have the top quarter being clipped and therefore a less useful test point.

remia replied:
Actually because this is not a full inversion from PQ, I think it actually needs to be scaled up to 10.0 (which would produce 1000 nits as 1.0 is 100 nits).

remia commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -347,6 +349,1305 @@ void Add_GamutComp_13_Inv_Shader(GpuShaderText & ss,
+    std::string findGamutBoundaryIntersectionName = _Add_Find_Gamut_Boundary_Intersection_func(shaderCreator, resourceIndex, g, solveJIntersectName);
+    std::string getReachBoundaryName = _Add_Reach_Boundary_func(shaderCreator, resourceIndex, g, reachName, getFocusGainName, solveJIntersectName);
+    std::string compressionName = _Add_Compression_func(shaderCreator, resourceIndex, true);
+    std::string gamutCompressName = _Add_Compress_Gamut_func(shaderCreator, resourceIndex, g, cuspName, getFocusGainName, gammaName, findGamutBoundaryIntersectionName, getReachBoundaryName, compressionName);
```
I'll add a test for this, it works though when you try to have 3 transforms in the same shader it starts to cause issues on my laptop, which I think is due to the const array we added for the hues lookup table that gets added for each transform. Seems like something that should go on the list for 2.4.1.

remia replied:
And this is specific to some OpenGL implmentation / hardware, on NVidia / Linux I can stack more transforms in a single shader without issues (tested 6 times).

remia commented on tests/gpu/FixedFunctionOp_test.cpp:
```diff
@@ -305,6 +305,463 @@ OCIO_ADD_GPU_TEST(FixedFunction, style_aces_gamutcomp13_inv)
+        1.041565657f, 1.041566014f, 1.041565657f, 1.0f,
+    };
+    test.setCustomValues(values);
+
```
Agreed, though on this test when I try it, it no longer pass with the current threshold which means I would have to increase. Think that's why I didn't use it so far.

doug-walker replied:
The formula for the relative comparison is:
`err = abs(x1 - x2) / max(abs(x1), min_val)`

Where min_val is set via setExpectedMinimalValue and defaults to 1e-6, which may be too small. I often set it to 1, or at least 1e-3. That's when it tends to allow reducing the tolerance. But the dynamic range of the values in the test is not very extreme, so perhaps it doesn't matter.

remia replied:
Thanks, I was missing this parameter, still on this particular test it doesn't allow to reduce the threshold much further, as you say the dynamic range is not very high. Will keep in mind for future works.

remia commented on src/OpenColorIO/ops/fixedfunction/ACES2/Transform.cpp:
```diff
@@ -0,0 +1,928 @@
+{
+    const float F_L_v = powf(F_L * std::abs(v) / reference_luminance, 0.42f);
+    return (400.f * std::copysign(1.f, v) * F_L_v) / (27.13f + F_L_v);
+}
```
Not sure either if it matters, I would tend to think it doesn't, adding a comment / todo for the time being.

remia commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -347,6 +349,1305 @@ void Add_GamutComp_13_Inv_Shader(GpuShaderText & ss,
+    ss.newLine() << "hwrap = hwrap - floor(hwrap / 360.f) * 360.f;";
+    ss.newLine() << "hwrap = (hwrap < 0.0f) ? hwrap + 360.f : hwrap;";
+
+    ss.newLine() << ss.floatDecl("i_lo") << " = floor(hwrap);";
```
I'm following the CTL code here but I don't remember exactly why the ReachM table doesn't have the additional 2 entries wrap like the Gamut cusp and the Gamma hull. @nick-shaw probably knows.

nick-shaw replied:
That table is evenly spaced in `h`, unlike the other cusp tables, and so the lookup code is different. I seem to remember that @scottdyer had a problem making the wrap-around work in one case. Perhaps he never worked out what his issue with that one was, so left it as is. The extra entries are really an implementation efficiency, not a core part of the algorithm.

remia commented:
I did some profiling on the CPU paths and GPU timings and didn't find regression since earlier results.

doug-walker reviewed (approved):
Thanks for the updates!

doug-walker commented:
Hi Remi, do you have any additional commits planned?  When you're done, please go ahead and merge it.  :)

remia commented:
Good to merge, I updated the branch against main, feel free to do it when the tests are completed!

KelSolaar commented:
Fantastic job Remi!

num3ric commented on src/OpenColorIO/ops/fixedfunction/FixedFunctionOpGPU.cpp:
```diff
@@ -347,6 +349,1305 @@ void Add_GamutComp_13_Inv_Shader(GpuShaderText & ss,
+    shaderCreator->addTexture(
+        name.c_str(),
+        GpuShaderText::getSamplerName(name).c_str(),
+        g.gamut_cusp_table.total_size,
```
Minor todo fix for later: same value, but uses `gamut_cusp_table` instead of `upper_hull_gamma_table`.
